### PR TITLE
Ansible: Backend parts round 2

### PR DESCRIPTION
**Pull request by hustodemon**
## What does this PR change?

(don't mind the branch name, it's wrong)

- Backend: refactoring of ansible code (moving to separate classes)
- Backend + XMLRPC + Controller: Fetching a single playbook contents
- Controller: Expose scheduling playbook execution
- Backend + XMLRPC + Controller: Discovery of playbooks in playbook path
- Backend + XMLRPC + Controller: Introspection of inventory path

Example XMLRPC output:

playbooks
```python
client.ansible.discoverPlaybooks(key, PATH_ID)

> {'/root/ansible-examples/lamp_haproxy': {
      'aws/demo-aws-launch.yml': {   'fullpath': '/root/ansible-examples/lamp_haproxy/aws/demo-aws-launch.yml'},
      'aws/rolling_update.yml': {   'fullpath': '/root/ansible-examples/lamp_haproxy/aws/rolling_update.yml'},
      'aws/site.yml': {   'fullpath': '/root/ansible-examples/lamp_haproxy/aws/site.yml'},
      'provision.yml': {   'custom_inventory': '/root/ansible-examples/lamp_haproxy/hosts',
      'fullpath': '/root/ansible-examples/lamp_haproxy/provision.yml'},
      'rolling_update.yml': {   'custom_inventory': '/root/ansible-examples/lamp_haproxy/hosts',
      'fullpath': '/root/ansible-examples/lamp_haproxy/rolling_update.yml'},
      'site.yml': {   'custom_inventory': '/root/ansible-examples/lamp_haproxy/hosts',
      'fullpath': '/root/ansible-examples/lamp_haproxy/site.yml'}}}
```

inventory
```python
client.ansible.introspectInventory(key, PATH_ID)

> {   '_meta': {   'hostvars': {   'web2': {   'dbname': 'foodb',
                                             'dbuser': 'foouser',
                                             'httpd_port': 80.0,
                                             'mysql_port': 3306.0,
                                             'mysqlservice': 'mysqld',
                                             'ntpserver': '192.168.1.2',
                                             'repository': 'https://github.com/bennojoy/mywebapp.git',
                                             'upassword': 'abc'},
                                 'web3': {   'httpd_port': 80.0,
                                             'ntpserver': '192.168.1.2',
                                             'repository': 'https://github.com/bennojoy/mywebapp.git'}}},

    'all': {   'children': ['dbservers', 'ungrouped', 'webservers']},
    'dbservers': {   'hosts': ['web2']},
    'webservers': {   'hosts': ['web3']}}
```

playbook contents

```python
ansible.fetchPlaybookContents(key, PATH_ID, "lamp_simple/site.yml")
>
---
# This playbook deploys the whole application stack in this site.

- name: apply common configuration to all nodes
  hosts: all
  remote_user: root

  roles:
    - common

- name: configure and deploy the webservers and application code
  hosts: webservers
  remote_user: root

  roles:
    - web

- name: deploy MySQL and configure the databases
  hosts: dbservers
  remote_user: root

  roles:
    - db
```

## GUI diff

No difference.

- [x] **DONE**

## Documentation
https://github.com/SUSE/spacewalk/issues/14571

- [x] **DONE**

## Test coverage
- Unit tests were added

- [x] **DONE**

## Links
https://github.com/SUSE/spacewalk/issues/14533
https://github.com/SUSE/spacewalk/issues/14518
https://github.com/SUSE/spacewalk/issues/14517
https://github.com/SUSE/spacewalk/issues/14525
https://github.com/SUSE/spacewalk/issues/14526

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/redhat/rhn/manager/system/test/AnsibleManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+     * @throws Exception
+     */
+    public void testSaveAndLookupAnsiblePath() throws Exception {
+        MinionServer minion = createAnsibleControlNode(user);
+        AnsiblePath path = new InventoryPath(minion);
+        path.setPath(Path.of("/tmp/test1"));
+        path = AnsibleManager.createAnsiblePath("inventory", minion.getId(), "/tmp/test", user);
+
+        HibernateFactory.getSession().flush();
+        HibernateFactory.getSession().evict( path);
```
Comment (cbbayburt) [inline java/code/src/com/redhat/rhn/manager/system/test/AnsibleManagerTest.java:86]:
```suggestion
        HibernateFactory.getSession().evict(path);
```

File: java/code/src/com/suse/manager/webui/utils/gson/AnsiblePlaybookExecutionJson.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+import org.apache.commons.lang3.builder.ToStringBuilder;
+
+import java.util.Date;
+import java.util.Optional;
+
+/**
+ * DTO for identifying ansible playbook execution
+ */
+public class AnsiblePlaybookExecutionJson {
```
Comment (cbbayburt) [inline java/code/src/com/suse/manager/webui/utils/gson/AnsiblePlaybookExecutionJson.java:26]: We need to support action chains with an optional parameter

File: java/code/src/com/suse/manager/webui/utils/gson/AnsiblePlaybookExecutionJson.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this)
+                .append("playbookPath", playbookPath)
+                .append("inventoryPath", inventoryPath)
+                .append("controlNodeId", controlNodeId)
+                .append("earliest", earliest)
+                .toString();
+    }
```
Comment (cbbayburt) [inline java/code/src/com/suse/manager/webui/utils/gson/AnsiblePlaybookExecutionJson.java:77]: Not so important, but do the `Optional`s work fine here?

**Review by cbbayburt — CHANGES_REQUESTED** [review]
Looks great, @hustodemon :+1: Great test coverage.

The only important thing from my side is the action chains support (commented below)

File: java/code/src/com/suse/manager/webui/utils/gson/AnsiblePlaybookExecutionJson.java (hunk 2 above)
Comment (hustodemon) [reply java/code/src/com/suse/manager/webui/utils/gson/AnsiblePlaybookExecutionJson.java:26]: Done. Thanks for the reminder!

**Review by cbbayburt — APPROVED** [review]

**Comment by hustodemon** [comment]
Merged into the feature branch. @cbbayburt thanks for your review!

**Pull request merged**